rialto_airflow/harvest/deduplicate.py

```diff
@@ -31,19 +44,63 @@ def remove_wos_duplicates(snapshot: Snapshot) -> int:
-            main_pub = pubs[0].id
-            for pub in pubs[1:]:
-                # Move author relationships to the first instance
-                for author in pub.authors:
-                    session.execute(
-                        insert(pub_author_association).values(
-                            publication_id=main_pub, author_id=author.id
-                        )
+            deleted = merge_pubs(pubs=pubs, session=session)
+            count_deleted += deleted
+        logging.info(f"Deleted a total of {count_deleted} publication rows from WOS.")
+    return num_dupes
+
+
+def remove_openalex_duplicates(snapshot: Snapshot) -> int:
+    logging.info("Removing any duplicate OpenAlex publications.")
+    with get_session(snapshot.database_name).begin() as session:
+        # Find all duplicate OpenAlex publications in the snapshot
+        duplicates = session.execute(
+            select(func.count(), Publication.openalex_json["id"])
+            .where(Publication.doi.is_(None))
+            .where(Publication.openalex_json["id"].is_not(None))
+            .group_by(Publication.openalex_json["id"])
+            .having(func.count() > 1)
+        ).all()
+        num_dupes = len(duplicates)
+        logging.info(f"Found {num_dupes} publications with duplicates.")
+        count_deleted = 0
+        openalex_ids = [row[1] for row in duplicates]
+        for openalex_id in openalex_ids:
+            pubs = (
+                session.execute(
+                    select(Publication).where(
+                        Publication.openalex_json["id"].astext == openalex_id
                     )
-                # Delete the duplicate
-                session.execute(delete(Publication).where(Publication.id == pub.id))  # type: ignore
-                count_deleted += 1
-        logging.info(f"Deleted a total of {count_deleted} publication rows.")
+                )
+                .scalars()
+                .all()
+            )
+            deleted = merge_pubs(pubs=pubs, session=session)
+            count_deleted += deleted
+        logging.info(
+            f"Deleted a total of {count_deleted} publication rows from OpenAlex."
+        )
     return num_dupes
+
+
+def merge_pubs(*, pubs, session) -> int:
+    """
+    Given a set of publications that are duplicates, merge them into one,
+    moving author relationships to the first instance and deleting the duplicates.
+    Returns the number of deleted publications.
+    """
+    count_deleted = 0
+    main_pub = pubs[0].id
+    for pub in pubs[1:]:
+        # Move author relationships to the first instance
+        for author in pub.authors:
+            session.execute(
+                insert(pub_author_association)
+                .values(publication_id=main_pub, author_id=author.id)
+                .on_conflict_do_nothing()
+            )
+
+        # Delete the duplicate
+        session.execute(delete(Publication).where(Publication.id == pub.id))  # type: ignore
```

Comment:
do/should these happen in a transaction?

Reply:
These changes are part of the session that's passed into the function, and which is held open with the context manager `with get_session(snapshot.database_name).begin() as session:`

The context manager should handle committing or rolling back if there are any errors. See: https://docs.sqlalchemy.org/en/14/orm/session_basics.html#framing-out-a-begin-commit-rollback-block 

So I think this is covered? But possible there's something I don't understand with sqlalchemy.

Reply:
sounds good, my comment should not be taken as knowing something more than that !

Reply:
I'm going to test this in stage all the same.